File: apps/web/lib/data/coverage.ts
import {
  coverageRequestState,
  coverageSortKey,
  isActionRequired,
  isMissedCoverageIncident,
  outcomeLabel,
  rungDeadlineIso,
  rungLabel,
  type CoverageOutcome,
  type CoverageRequestInput,
  type CoverageRequestState,
  type CoverageRung,
} from '@shift/core';
import { cache } from 'react';

import { getSessionUser } from '../auth';
import { cachedGlobal } from '../cache/ttl';
import { createClient, createServiceClient } from '../supabase/server';

// ===========================================================================
// Allied coverage requests — READ model.
//
// RLS on allied_coverage_requests scopes rows to user_can_build_schedule(house) plus
// an admin clause, so the AUTHED client is correct here (no service bypass), matching
// the Action Inbox read path.
//
// All lifecycle derivation lives in @shift/core (`coverageRequestState`). This module
// only does I/O and formatting. The central rule it surfaces: an OPEN request never
// auto-clears. Once its window passes it is `overdue` and stays on screen until a
// human records an outcome.
// ===========================================================================

const NY = 'America/New_York';

export type CoverageItem = {
  id: string;
  houseId: string;
  houseName: string;
  state: CoverageRequestState;
  actionRequired: boolean;
  isIncident: boolean;
  dateLabel: string; // "Tue, Jun 24"
  windowLabel: string; // "22:00 to 23:00"
  windowStartIso: string;
  windowEndIso: string;
  reason: string;
  rung: CoverageRung;
  rungLabel: string;
  rungDeadlineIso: string | null;
  recipientName: string | null;
  acknowledgedByName: string | null;
  outcome: CoverageOutcome | null;
  outcomeLabel: string | null;
  closeNote: string | null;
  closedByName: string | null;
  closedAtIso: string | null;
};

export type CoverageData = {
  overdue: CoverageItem[];
  awaitingAck: CoverageItem[];
  acknowledged: CoverageItem[];
  actionRequiredCount: number; // drives the banner and the red bell badge
  openCount: number;
};

const REASON: Record<string, string> = {
  float_no_acknowledgment: 'No floater found or the floater did not acknowledge in time.',
  no_floater_found: 'No floater found in the eligible source houses.',
  floater_declined: 'The assigned floater declined.',
  escalation_chain: 'The desk will be empty and no one picked up the shift.',
};

function prettifyHouse(id: string): string {
  const m = /^house-(\d+)$/.exec(id);
  if (m) return `House ${String(Number(m[1]))}`;
  return id.charAt(0).toUpperCase() + id.slice(1);
}

function dayDateLabel(iso: string): string {
  return new Intl.DateTimeFormat('en-US', {
    timeZone: NY,
    weekday: 'short',
    month: 'short',
    day: 'numeric',
  }).format(new Date(iso));
}

function hm(iso: string): string {
  return new Intl.DateTimeFormat('en-GB', {
    timeZone: NY,
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  }).format(new Date(iso));
}

type CoverageRow = {
  request_id: string;
  house_id: string;
  window_start_at: string;
  window_end_at: string;
  reason: string;
  current_rung: string;
  rung_fired_at: string;
  current_recipient: string | null;
  acknowledged_at: string | null;
  acknowledged_by: string | null;
  closed_at: string | null;
  closed_by: string | null;
  outcome: CoverageOutcome | null;
  close_note: string | null;
};

function toInput(row: CoverageRow): CoverageRequestInput {
  return {
    windowStartIso: row.window_start_at,
    windowEndIso: row.window_end_at,
    acknowledgedAtIso: row.acknowledged_at,
    closedAtIso: row.closed_at,
    outcome: row.outcome,
    currentRung: row.current_rung as CoverageRung,
    rungFiredAtIso: row.rung_fired_at,
  };
}

const SELECT =
  'request_id, house_id, window_start_at, window_end_at, reason, current_rung, ' +
  'rung_fired_at, current_recipient, acknowledged_at, acknowledged_by, closed_at, ' +
  'closed_by, outcome, close_note';

// The rung timeout is a system_config row, so the countdown must read it rather than
// assume the 60-minute default.
//
// Memoized process-wide (same rationale as the other global config reads in
// lib/cache/ttl): it is one operator-set row, identical for every caller and not
// derived from the signed-in user, but it sat on the critical path of every coverage
// render as its own remote round trip (~150ms against the hosted project).
const RUNG_TIMEOUT_KEY = 'system_config:allied_ladder_rung_timeout_minutes';
const RUNG_TIMEOUT_TTL_MS = 60_000;

async function rungTimeoutMinutes(): Promise<number> {
  return cachedGlobal(RUNG_TIMEOUT_KEY, RUNG_TIMEOUT_TTL_MS, async () => {
    const svc = createServiceClient();
    const { data } = await svc
      .from('system_config')
      .select('config_value')
      .eq('config_key', 'allied_ladder_rung_timeout_minutes')
      .maybeSingle();
    const parsed = Number(data?.config_value);
    return Number.isInteger(parsed) && parsed > 0 ? parsed : 60;
  });
}

// Names for the people on the ladder. A separate lookup because
// allied_coverage_requests has no FK-embeddable relationship the authed client may
// traverse under RLS on `users`.
async function nameMap(ids: (string | null)[]): Promise<Map<string, string>> {
  const unique = [...new Set(ids.filter((id): id is string => id !== null))];
  if (unique.length === 0) return new Map();
  const svc = createServiceClient();
  const { data } = await svc.from('users').select('user_id, name').in('user_id', unique);
  return new Map((data ?? []).map((u) => [u.user_id, u.name]));
}

function enrich(
  row: CoverageRow,
  nowIso: string,
  timeoutMinutes: number,
  names: Map<string, string>,
): CoverageItem {
  const input = toInput(row);
  return {
    id: row.request_id,
    houseId: row.house_id,
    houseName: prettifyHouse(row.house_id),
    state: coverageRequestState(input, nowIso),
    actionRequired: isActionRequired(input, nowIso),
    isIncident: isMissedCoverageIncident(input, nowIso),
    dateLabel: dayDateLabel(row.window_start_at),
    // "to", not a dash: no em dash or en dash in user-facing copy.
    windowLabel: `${hm(row.window_start_at)} to ${hm(row.window_end_at)}`,
    windowStartIso: row.window_start_at,
    windowEndIso: row.window_end_at,
    reason: REASON[row.reason] ?? row.reason.replace(/_/g, ' '),
    rung: row.current_rung as CoverageRung,
    rungLabel: rungLabel(row.current_rung as CoverageRung),
    rungDeadlineIso: rungDeadlineIso(input, timeoutMinutes),
    recipientName: row.current_recipient ? (names.get(row.current_recipient) ?? null) : null,
    acknowledgedByName: row.acknowledged_by ? (names.get(row.acknowledged_by) ?? null) : null,
    outcome: row.outcome,
    outcomeLabel: row.outcome ? outcomeLabel(row.outcome) : null,
    closeNote: row.close_note,
    closedByName: row.closed_by ? (names.get(row.closed_by) ?? null) : null,
    closedAtIso: row.closed_at,
  };
}

// Every OPEN request the signed-in manager may see, grouped by urgency.
//
// Wrapped in React's cache() (per-request), because the shell layout AND the /inbox
// page both need this on the same render: the layout for the app-wide banner and the
// bell badge, the page for the Coverage tab. Unmemoized, /inbox paid this whole read
// TWICE — the requests select, the rung-timeout config row, and the name lookup — for
// one navigation. The cache key is the `now` argument, and simNow() is itself cache()d
// so the layout and the page pass the SAME Date instance and share the entry.
//
// Per-request, not process-wide: the rows are RLS-scoped to the signed-in manager, so a
// cross-request cache here would leak one house's coverage to another's.
export const getCoverageData = cache(async (now: Date = new Date()): Promise<CoverageData> => {
  const empty: CoverageData = {
    overdue: [],
    awaitingAck: [],
    acknowledged: [],
    actionRequiredCount: 0,
    openCount: 0,
  };

  const supabase = await createClient();
  // getSessionUser() rather than supabase.auth.getUser(): the latter is an HTTP call to
  // GoTrue on every invocation, and this function only needs to know whether anyone is
  // signed in. getSessionUser() verifies the token locally and is already resolved for
  // this request by the layout, so this costs nothing (see the note on it in lib/auth).
  const user = await getSessionUser();
  if (user === null) return empty;

  const { data: rows, error } = await supabase
    .from('allied_coverage_requests')
    .select(SELECT)
    .is('closed_at', null)
    .order('window_start_at', { ascending: true })
    .limit(200);

  // NEVER swallow this. A failed read here renders as "All clear. No coverage needed",
  // which is the single most dangerous lie this app can tell: it is indistinguishable
  // from a genuinely quiet night while a desk is about to go unstaffed. A missing table
  // GRANT produced exactly that during development on 2026-07-29.
  if (error !== null) {
    console.error(
      JSON.stringify({ event: 'coverage_read_failed', message: error.message, code: error.code }),
    );
    throw new Error(`Coverage requests could not be loaded: ${error.message}`);
  }

  // `as unknown as` because the select string is built from a const, which PostgREST's
  // generic cannot narrow to the row shape on its own.
  const all = (rows ?? []) as unknown as CoverageRow[];
  if (all.length === 0) return empty;

  const nowIso = now.toISOString();
  const [timeoutMinutes, names] = await Promise.all([
    rungTimeoutMinutes(),
    nameMap(all.flatMap((r) => [r.current_recipient, r.acknowledged_by])),
  ]);

  // Overdue first (most overdue at the top), then soonest window. The key is computed
  // once per row alongside the enrich, not looked back up during the comparator.
  const items = all
    .map((r) => ({
      item: enrich(r, nowIso, timeoutMinutes, names),
      key: coverageSortKey(toInput(r), nowIso),
    }))
    .sort((a, b) => a.key - b.key)
    .map((entry) => entry.item);

  return {
    overdue: items.filter((i) => i.state === 'overdue'),
    awaitingAck: items.filter((i) => i.state === 'awaiting_ack'),
    acknowledged: items.filter((i) => i.state === 'acknowledged'),
    actionRequiredCount: items.filter((i) => i.actionRequired).length,
    openCount: items.length,
  };
});

// The shell's read of the same data. The throw above is correct for /inbox, which has an
// error.tsx to catch it, but NOT for (app)/layout.tsx: a layout renders ABOVE every
// error boundary in its own subtree, so an uncaught throw there escapes to Next's
// built-in global error. In production that is a blank error screen for the WHOLE admin
// console on every route; in dev the error overlay reloads the document, the layout
// throws again on the reload, and the app reload-loops about twice a second with no way
// to navigate out. That is exactly what a stale PostgREST schema cache for
// allied_coverage_requests produced on 2026-07-29.
//
// So the shell degrades instead of crashing, and it still refuses to lie: `unavailable`
// makes the banner say the coverage status could not be read, never "all clear".
export async function getShellCoverage(
  now: Date,
): Promise<{ data: CoverageData | null; unavailable: boolean }> {
  try {
    return { data: await getCoverageData(now), unavailable: false };
  } catch (err) {
    console.error(
      JSON.stringify({
        event: 'shell_coverage_read_failed',
        message: err instanceof Error ? err.message : String(err),
      }),
    );
    return { data: null, unavailable: true };
  }
}

// The single number the app-wide banner and the red bell badge need. Kept separate and
// deliberately cheap: the shell layout renders it on EVERY page.
export async function getCoverageActionCount(now: Date = new Date()): Promise<number> {
  const supabase = await createClient();
  const user = await getSessionUser();
  if (user === null) return 0;

  const { data: rows } = await supabase
    .from('allied_coverage_requests')
    .select('window_end_at, acknowledged_at, closed_at')
    .is('closed_at', null)
    .limit(200);

  const nowIso = now.toISOString();
  return (rows ?? []).filter((r) =>
    isActionRequired(
      {
        windowStartIso: nowIso,
        windowEndIso: r.window_end_at as string,
        acknowledgedAtIso: r.acknowledged_at as string | null,
        closedAtIso: r.closed_at as string | null,
        outcome: null,
        currentRung: 'rsm',
        rungFiredAtIso: nowIso,
      },
      nowIso,
    ),
  ).length;
}

export type CoverageReportRow = CoverageItem & { incidentReason: string | null };

// The missed-coverage report (/admin/coverage). Includes CLOSED requests, which the
// live list deliberately excludes: this is the audit trail that the old
// archive-then-discard model destroyed.
export async function getCoverageReport(
  fromIso: string,
  toIso: string,
  now: Date = new Date(),
): Promise<CoverageReportRow[]> {
  const supabase = await createClient();
  const user = await getSessionUser();
  if (user === null) return [];

  const { data: rows } = await supabase
    .from('allied_coverage_requests')
    .select(SELECT)
    .gte('window_start_at', fromIso)
    .lte('window_start_at', toIso)
    .order('window_start_at', { ascending: false })
    .limit(500);

  const all = (rows ?? []) as unknown as CoverageRow[];
  if (all.length === 0) return [];

  const nowIso = now.toISOString();
  const [timeoutMinutes, names] = await Promise.all([
    rungTimeoutMinutes(),
    nameMap(all.flatMap((r) => [r.current_recipient, r.acknowledged_by, r.closed_by])),
  ]);

  return all.map((r) => {
    const item = enrich(r, nowIso, timeoutMinutes, names);
    return {
      ...item,
      incidentReason: item.isIncident
        ? r.outcome === 'desk_unstaffed'
          ? 'The desk went unstaffed.'
          : 'The coverage window passed with no close-out.'
        : null,
    };
  });
}

File: apps/mobile/shared/src/commonMain/kotlin/com/pennhousing/shift/shared/manager/coverage/Coverage.kt
package com.pennhousing.shift.shared.manager.coverage

import com.pennhousing.shift.shared.shifts.NEW_YORK
import com.pennhousing.shift.shared.shifts.formatBlockTime
import com.pennhousing.shift.shared.shifts.formatDayLabel
import com.pennhousing.shift.shared.shifts.formatDuration
import kotlinx.datetime.TimeZone
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Instant

/*
 * Allied coverage-request lifecycle — PURE presentation + state logic for the manager
 * Coverage tab (BSpec §5.4a). No I/O, no system clock: `now` is injected, per the
 * phase-13a rule that tested logic never reads a clock.
 *
 * CROSS-PLATFORM PARITY. This is a KOTLIN MIRROR of `packages/core/src/coverage/index.ts`,
 * which drives the web Coverage page. The two must agree on state derivation, the
 * action-required rule, the sort order, the rung labels, the outcome labels and the
 * note requirement, or the same request reads differently on a manager's phone than on
 * their laptop. `CoverageTest` pins this copy against the TS one's behaviour. If you
 * change either, change both. This is the same class of hazard as
 * `house/WorkerColors.kt` vs `apps/web/lib/workerColor.ts`.
 *
 * THE CENTRAL RULE, restated from the TS module because it is easy to undo:
 *
 *     AN OPEN REQUEST NEVER AUTO-CLEARS.
 *
 * Once the coverage window passes without a close-out the request is `OVERDUE` and stays
 * on screen until a human records an outcome. `CLOSED` is reachable only through
 * `close_allied_coverage_request`. There is no discard.
 *
 *   AWAITING_ACK -> ACKNOWLEDGED -> CLOSED     (the healthy path)
 *   AWAITING_ACK -> OVERDUE                    (window passed, nobody acknowledged)
 *   ACKNOWLEDGED -> OVERDUE                    (acknowledged, window passed, not closed)
 *
 * A NOTE ON THE MERGED UI. The TS module's header warns that collapsing "acknowledge"
 * and "close out" into one control is what lost the audit trail in the predecessor
 * design. The mobile Respond sheet (docs/manager-app/SPEC.md §6.1) does present them as
 * one job to the manager, and that is a deliberate product decision, NOT a reversal of
 * that warning: the two STATES remain distinct here and in the database. What is merged
 * is the sequence of taps, not the record. Acknowledgement still happens the moment the
 * sheet opens (so the ladder stops immediately) and the outcome is still recorded
 * separately, because it is not known until the Allied call connects. Do not "simplify"
 * this by dropping ACKNOWLEDGED.
 */

/** Why the escalation chain ran out of internal options. `reason` on the request row. */
private val REASON_LABEL =
    mapOf(
        "float_no_acknowledgment" to "No floater found or the floater did not acknowledge in time.",
        "no_floater_found" to "No floater found in the eligible source houses.",
        "floater_declined" to "The assigned floater declined.",
        "escalation_chain" to "The desk will be empty and no one picked up the shift.",
    )

/** Human copy for a raw `reason`. Unknown values degrade to the de-underscored raw value. */
fun coverageReasonLabel(reason: String): String = REASON_LABEL[reason] ?: reason.replace('_', ' ')

/** The four ways a request may be closed (`allied_coverage_outcome`). */
enum class CoverageOutcome(val wire: String) {
    ALLIED_SECURED("allied_secured"),
    COVERED_INTERNALLY("covered_internally"),
    DESK_UNSTAFFED("desk_unstaffed"),
    NO_LONGER_NEEDED("no_longer_needed"),
    ;

    companion object {
        fun fromWire(wire: String?): CoverageOutcome? = entries.firstOrNull { it.wire == wire }
    }
}

/**
 * A note is mandatory when reporting that a desk went empty. Mirrors the RPC's
 * `note_required` guard so the sheet can disable the button before the round trip; the
 * RPC stays authoritative. An unexplained `desk_unstaffed` is the row nobody can act on
 * later.
 */
fun requiresCloseNote(outcome: CoverageOutcome): Boolean = outcome == CoverageOutcome.DESK_UNSTAFFED

/** Display copy for an outcome. No em/en dashes: this is surfaced text. */
fun outcomeLabel(outcome: CoverageOutcome): String =
    when (outcome) {
        CoverageOutcome.ALLIED_SECURED -> "Allied secured"
        CoverageOutcome.COVERED_INTERNALLY -> "Covered internally"
        CoverageOutcome.DESK_UNSTAFFED -> "Desk went unstaffed"
        CoverageOutcome.NO_LONGER_NEEDED -> "No longer needed"
    }

/** The three ladder rungs, plus the project-administrator terminal (BSpec §2.6). */
enum class CoverageRung(val wire: String) {
    RSM("rsm"),
    HM("hm"),
    HMOD("hmod"),
    ADMIN("admin"),
    ;

    /**
     * Nobody sits above the HMOD or the project administrator, so the request stays put
     * and keeps reminding. The UI says so explicitly rather than showing a countdown to
     * an escalation that will never happen.
     */
    val isTerminal: Boolean get() = this == HMOD || this == ADMIN

    companion object {
        /** Unknown rung values resolve to the terminal rung: never promise an escalation we cannot make. */
        fun fromWire(wire: String?): CoverageRung = entries.firstOrNull { it.wire == wire } ?: HMOD
    }
}

fun rungLabel(rung: CoverageRung): String =
    when (rung) {
        CoverageRung.RSM -> "Residential Services Manager"
        CoverageRung.HM -> "Housing Manager"
        CoverageRung.HMOD -> "Housing Manager on duty"
        CoverageRung.ADMIN -> "Project administrator"
    }

/** Where a request sits right now. */
enum class CoverageRequestState { AWAITING_ACK, ACKNOWLEDGED, OVERDUE, CLOSED }

/**
 * One `allied_coverage_requests` row, as the repository read it. Times are [Instant], so
 * ordering is real time-line ordering: the TS mirror has a comment warning that
 * offset-bearing ISO strings do not sort lexically, and parsing to [Instant] here is how
 * this copy avoids that trap entirely.
 */
data class CoverageRequest(
    val requestId: String,
    val houseId: String,
    val houseName: String,
    val windowStart: Instant,
    val windowEnd: Instant,
    val reason: String,
    val currentRung: CoverageRung,
    val rungFiredAt: Instant,
    val acknowledgedAt: Instant? = null,
    val closedAt: Instant? = null,
    val outcome: CoverageOutcome? = null,
    /** The desk phone for this house, when known. Shown on the Respond sheet's call action. */
    val deskPhone: String? = null,
)

/**
 * Where a request sits at [now].
 *
 * The ORDER OF THESE CHECKS IS LOAD-BEARING and matches the TS mirror: `CLOSED` wins over
 * everything (a closed request is finished even if its window is long past), and `OVERDUE`
 * wins over `ACKNOWLEDGED` (an acknowledged request whose window elapsed without a
 * close-out is exactly the case that must stay visible). Reordering these silently hides
 * missed coverage.
 */
fun coverageRequestState(
    request: CoverageRequest,
    now: Instant,
): CoverageRequestState =
    when {
        request.closedAt != null -> CoverageRequestState.CLOSED
        now >= request.windowEnd -> CoverageRequestState.OVERDUE
        request.acknowledgedAt != null -> CoverageRequestState.ACKNOWLEDGED
        else -> CoverageRequestState.AWAITING_ACK
    }

/**
 * Is this request still demanding someone's attention? Drives the app-wide banner and the
 * Coverage tab badge. An acknowledged request is deliberately NOT action-required:
 * somebody has said they are handling it, and the banner must stop nagging them.
 */
fun isActionRequired(
    request: CoverageRequest,
    now: Instant,
): Boolean =
    when (coverageRequestState(request, now)) {
        CoverageRequestState.AWAITING_ACK, CoverageRequestState.OVERDUE -> true
        else -> false
    }

/**
 * A missed-coverage INCIDENT: either a desk that went unstaffed, or a request nobody ever
 * closed. Both mean the process failed. The mobile app does not render the incident report
 * (SPEC §6.6 keeps it on web), but the predicate lives here so the card can mark itself.
 */
fun isMissedCoverageIncident(
    request: CoverageRequest,
    now: Instant,
): Boolean =
    if (request.closedAt != null) {
        request.outcome == CoverageOutcome.DESK_UNSTAFFED
    } else {
        coverageRequestState(request, now) == CoverageRequestState.OVERDUE
    }

/**
 * When the current rung escalates, given the configured timeout. Null once the request is
 * acknowledged or closed (the ladder has stopped) or on a terminal rung (nobody above).
 */
fun rungDeadline(
    request: CoverageRequest,
    timeoutMinutes: Int,
): Instant? {
    if (request.acknowledgedAt != null || request.closedAt != null) return null
    if (request.currentRung.isTerminal) return null
    return request.rungFiredAt + timeoutMinutes.minutes
}

/**
 * "Escalates in 12m" / "Escalating now" / "No further escalation" — the live countdown on
 * the card. Null when the ladder has stopped because someone acknowledged.
 */
fun rungCountdownLabel(
    request: CoverageRequest,
    timeoutMinutes: Int,
    now: Instant,
): String? {
    if (request.acknowledgedAt != null || request.closedAt != null) return null
    if (request.currentRung.isTerminal) return "No further escalation"
    val deadline = rungDeadline(request, timeoutMinutes) ?: return null
    if (now >= deadline) return "Escalating now"
    val left = deadline - now
    val h = left.inWholeMinutes / 60
    val m = left.inWholeMinutes % 60
    return if (h > 0) "Escalates in ${h}h ${m}m" else "Escalates in ${m}m"
}

/**
 * "Wed, 22:00 to 00:00" — the TRUE coverage window, from the request row's own
 * `window_start_at` / `window_end_at`.
 *
 * Never reconstruct this as start + 30 minutes. Migration 20260729000010's header records
 * that exact regression: a dropped `block_end_at` payload key made every Allied alert
 * render as a 30-minute window and archive up to 3.5 hours early on a 4-hour gap. The
 * window lives on the request row precisely so it cannot silently fall back again.
 */
fun coverageWindowLabel(
    request: CoverageRequest,
    zone: TimeZone = NEW_YORK,
): String =
    "${formatDayLabel(request.windowStart, zone)}, " +
        "${formatBlockTime(request.windowStart, zone)} to ${formatBlockTime(request.windowEnd, zone)}"

/** Whole hours of desk time at stake, for the card's "4h" chip. */
fun coverageWindowHours(request: CoverageRequest): Double =
    (request.windowEnd - request.windowStart).inWholeMinutes / 60.0

/**
 * Sort key for the Coverage list: overdue first (most overdue at the top), then the
 * soonest window. A manager scanning this list needs the thing that has already gone
 * wrong before the thing that is about to.
 *
 * The TS mirror subtracts `Number.MAX_SAFE_INTEGER` to force overdue rows ahead. This copy
 * returns a (bucket, key) pair instead, which is the same ordering without relying on a
 * sentinel that could collide at extreme timestamps.
 */
fun coverageSortKey(
    request: CoverageRequest,
    now: Instant,
): Pair<Int, Long> {
    val overdue = coverageRequestState(request, now) == CoverageRequestState.OVERDUE
    return if (overdue) {
        0 to request.windowEnd.toEpochMilliseconds()
    } else {
        1 to request.windowStart.toEpochMilliseconds()
    }
}

/** A fully-formatted Coverage card. The UI renders this directly and decides nothing. */
data class CoverageCard(
    val requestId: String,
    val houseId: String,
    val houseName: String,
    val windowLabel: String,
    val hoursLabel: String,
    val reasonLabel: String,
    val state: CoverageRequestState,
    /** "Housing Manager on duty" — who holds it now. */
    val rungLabel: String,
    /** "Escalates in 12m" / "No further escalation"; null once acknowledged. */
    val countdownLabel: String?,
    /** True on the terminal rung: the UI says nobody is coming after this. */
    val isTerminalRung: Boolean,
    val isActionRequired: Boolean,
    val isMissedCoverageIncident: Boolean,
    val deskPhone: String? = null,
    /** Set once closed, for the brief confirmation the card shows before it leaves the list. */
    val outcomeLabel: String? = null,
)

fun CoverageRequest.toCard(
    timeoutMinutes: Int,
    now: Instant,
    zone: TimeZone = NEW_YORK,
): CoverageCard {
    return CoverageCard(
        requestId = requestId,
        houseId = houseId,
        houseName = houseName,
        windowLabel = coverageWindowLabel(this, zone),
        // "4h" / "2h 30m" — the same duration formatter every other shift surface uses.
        hoursLabel = formatDuration(windowStart, windowEnd),
        reasonLabel = coverageReasonLabel(reason),
        state = coverageRequestState(this, now),
        rungLabel = rungLabel(currentRung),
        countdownLabel = rungCountdownLabel(this, timeoutMinutes, now),
        isTerminalRung = currentRung.isTerminal,
        isActionRequired = isActionRequired(this, now),
        isMissedCoverageIncident = isMissedCoverageIncident(this, now),
        deskPhone = deskPhone,
        outcomeLabel = outcome?.let { outcomeLabel(it) },
    )
}

/**
 * The Coverage tab's whole render state.
 *
 * Closed requests are absent by design (SPEC §6.1): the retrospective incident report
 * lives on web. [actionRequiredCount] is what the tab badge and the persistent banner
 * read; [openCount] includes acknowledged-but-not-closed requests, which still need a
 * human to record an outcome.
 */
data class CoverageFeed(
    val cards: List<CoverageCard>,
    val actionRequiredCount: Int,
    val openCount: Int,
) {
    val isEmpty: Boolean get() = cards.isEmpty()

    /**
     * Whether to show the non-dismissable app-wide banner: only for a request nobody has
     * picked up yet. Once acknowledged it downgrades to the tab badge, so a manager who
     * has said "I've got this" is not nagged on every screen while they are on the phone
     * to Allied.
     */
    val showsBanner: Boolean get() = actionRequiredCount > 0
}

/**
 * Build the Coverage feed: drop closed requests, order overdue-first then soonest-window,
 * and count what needs attention.
 */
fun buildCoverageFeed(
    requests: List<CoverageRequest>,
    timeoutMinutes: Int,
    now: Instant,
    zone: TimeZone = NEW_YORK,
): CoverageFeed {
    val open = requests.filter { coverageRequestState(it, now) != CoverageRequestState.CLOSED }
    val ordered =
        open.sortedWith(
            compareBy({ coverageSortKey(it, now).first }, { coverageSortKey(it, now).second }),
        )
    return CoverageFeed(
        cards = ordered.map { it.toCard(timeoutMinutes, now, zone) },
        actionRequiredCount = open.count { isActionRequired(it, now) },
        openCount = open.size,
    )
}

/** The system default for `allied_ladder_rung_timeout_minutes` (migration 20260729000010). */
const val DEFAULT_RUNG_TIMEOUT_MINUTES: Int = 60
